### PR TITLE
Make rusty-rlp optional

### PR DESCRIPTION
**Pull request by cburgdorf**
### What is wrong

We seem to have the following problem. Currently, we do not have 32-bit binaries for `rusty-rlp`. The reason for that is that Github Actions [does currently not support 32 bit platforms](https://github.community/t/does-github-actions-support-multiple-os-architectures/16526/2)

I had several people running into that problem. People will also run into this if they do have a 64 bit machine but installed a 32 bit Python [as it was the case here (solved)](https://github.com/cburgdorf/rusty-rlp/issues/15#issuecomment-691704680).

### How was it fixed

1. Only use `rusty-rlp` when it is explicitly installed e.g. `rlp[rust-backend]`

2. Add Python 3.8 support in CI

3. Add a job to test functionality with the rust backend

### CODE REVIEW

File: tox.ini
Comment (cburgdorf) [inline tox.ini:18]: That's probably not the supposed way of handling it but it works. Naturally I first tried `pip uninstall -y rusty-rlp && py.test {posargs:tests/}` but `tox` doesn't allow that. Then I noticed that the uninstall is already enough because (for some reason) it still derives the `py.test {posargs:tests/}` from `core` so it does what it is supposed to. I'm very open to learn about cleaner alternatives :D

File: rlp/codec.py
Comment (cburgdorf) [inline rlp/codec.py:29]: The warning can be observed by running `tox -e py38-core-fallback -- --log-cli-level WARNING`

I think it is especially useful to avoid situations where people are on a 64-bit system but have installed a 32-bit Python.

File: tox.ini
Comment (carver) [reply tox.ini:18]: Yeah, the prefixes match on substrings, so the `core: py.test {posargs:tests/}` line still matches `core-fallback` environments. I'm not clear on how ordering works, if it's even stable. My read of the config file is that pytest would run before the pip uninstall, but that's based on intuition, not a careful reading of the tox.ini spec.

Comment (cburgdorf) [reply tox.ini:18]: Ah that explains it. Regarding ordering. I don't know either but it appeared to work (I tested it with intentional making it fail and ultimately also with the log message that I introduced). I could also try to make it fail if the log message isn't in the output so that we can be sure it fails loudly if there actually is an issue with that approach.

File: setup.py
Comment (pipermerriam) [inline setup.py:58]: My spider sense says that logic like this doesn't work at installation time in certain contexts.  For example, I *think* that when installing with a wheel this file may not be evaluated?!???  but I'm really not sure about this.

**Review by pipermerriam — APPROVED** [review]

File: setup.py
Comment (cburgdorf) [reply setup.py:58]: I'm not sure I understand. You mean the fact of having the ternary expression at this particular place or evaluating the `is_64_bit` in general? If the former, then would moving the `install_requires` out of the `setup` function change that? We certainly [build it dynamically in Trinity, too](https://github.com/ethereum/trinity/blob/2aa13a83fce515502c78cf7f8013ba192816e607/setup.py#L121-L134)

If you mean the whole evaluation of the `is_64_bit` would not work then I wouldn't really know how to provide a proper fallback at all.

Comment (cburgdorf) [reply setup.py:58]: @pipermerriam Any update on this? I got two more requests [asking for ARM builds](https://github.com/cburgdorf/rusty-rlp/issues/17#issuecomment-701979389) (which I can not provide right now, too). The 64 bit check may not be sufficient to exclude these (since there are 64 bit ARM chips) but the principle for how to handle the fallback should be more or less the same.

Comment (pipermerriam) [reply setup.py:58]: I meant relying on having a python expression in the `setup.py` file be reliably evaluated at installation time.  I think that it is possible that for wheels, it may be evaluated at the time the wheel is created, but not during installation...  but I'm not sure about this.

Comment (cburgdorf) [reply setup.py:58]: >Avoids arbitrary code execution for installation. (Avoids setup.py)

Mmh..sounds like it. Then, maybe the best option would be to not depend on `rusty-rlp` at all but have pyrlp use it if it is importable. And then Trinity would list it as a dependency separately, hence activate it that way. Web3.py and friends would not use it unless it is explicitly installed by the user.

File: setup.py
Comment (cburgdorf) [inline setup.py:32]: Following the example of [`eth-hash`](https://github.com/ethereum/eth-hash/blob/d05c8ac0710e2dc16aae2c8a1e406ee7cbe78871/setup.py#L28-L34)

File: setup.py
Comment (cburgdorf) [reply setup.py]: @carver @pipermerriam Alright, I made the changes to make this optional. Trinity can add a dependency `rlp[rust-backend]` to get the performance benefit. Waiting for re-approval. If we merge this today or tomorrow I could potentially still cut a new release this week.

**Review by pipermerriam — APPROVED** [review]
Has this actually been tested on one of the effected systems?

**Comment by cburgdorf** [comment]
>Has this actually been tested on one of the effected systems?

No, but with this change and as soon as we cut a new release nobody will end up using the rusty-rlp backend by default. So everyone installing web3.py on whatever system will get the old pyrlp that they used to have. Except for Trinity where we add a dependency to `rlp[rust-backend]`. It still won't help Trinity users on a 32 bit system or on another unsupported OS but I think that is less of a problem compared to a bunch of web3.py folks being affected.

**Comment by cburgdorf** [comment]
If no one objects, I will cut version `2.0.0a2` tomorrow.

**Comment by carver** [comment]
Lol at the `2**32` file

**Pull request merged**
